nav2_motion_primitives/include/nav2_motion_primitives/motion_primitive.hpp

```diff
@@ -24,127 +24,137 @@
+  std::shared_ptr<nav2_robot::Robot> robot_;
+  std::unique_ptr<ActionServer> action_server_;
 
-    auto status = onRun(command);
+  void configure()
+  {
+    RCLCPP_INFO(node_->get_logger(), "Configuring ", primitive_name_.c_str());
 
-    if (status != nav2_tasks::TaskStatus::SUCCEEDED) {
-      return nav2_tasks::TaskStatus::FAILED;
-    }
+    robot_ = std::make_unique<nav2_robot::Robot>(
+      node_->get_node_base_interface(),
+      node_->get_node_topics_interface(),
+      node_->get_node_logging_interface(),
+      true);
 
-    ResultMsg result;
-    status = cycle(result);
-    task_server_->setResult(result);
+    action_server_ = std::make_unique<ActionServer>(node_, primitive_name_,
+        std::bind(&MotionPrimitive::execute, this, std::placeholders::_1));
+  }
 
-    return status;
+  void cleanup()
+  {
+    robot_.reset();
+    action_server_.reset();
   }
 
-protected:
-  nav2_tasks::TaskStatus cycle(ResultMsg & result)
+  void execute(const typename std::shared_ptr<GoalHandle> goal_handle)
   {
-    auto time_since_msg = std::chrono::system_clock::now();
-    auto start_time = std::chrono::system_clock::now();
-    auto current_time = std::chrono::system_clock::now();
+    RCLCPP_INFO(node_->get_logger(), "Attempting ", primitive_name_.c_str());
+
+    initialCheck(goal_handle);
+
+    // Log a message every second
+    auto timer = node_->create_wall_timer(1s,
+        [&]() {RCLCPP_INFO(node_->get_logger(), "%s running...", primitive_name_.c_str());});
 
-    auto status = nav2_tasks::TaskStatus::FAILED;
+    loop(goal_handle);
 
+    stopRobot();
+  }
+
+  void initialCheck(const typename std::shared_ptr<GoalHandle> goal_handle)
+  {
+    if (onRun(goal_handle->get_goal()) != Status::SUCCEEDED) {
+      RCLCPP_INFO(node_->get_logger(), "Initial checks failed for ", primitive_name_.c_str());
+      goal_handle->abort(std::make_shared<typename ActionT::Result>());
+    }
+  }
+
+  void loop(const typename std::shared_ptr<GoalHandle> goal_handle)
+  {
+    auto result = std::make_shared<typename ActionT::Result>();
     rclcpp::Rate loop_rate(10);
+
     while (rclcpp::ok()) {
-      if (task_server_->cancelRequested()) {
-        RCLCPP_INFO(node_->get_logger(), "%s cancelled", taskName_.c_str());
-        task_server_->setCanceled();
-        status = nav2_tasks::TaskStatus::CANCELED;
-        break;
+      if (goal_handle->is_canceling()) {
+        RCLCPP_INFO(node_->get_logger(), "Canceling", primitive_name_.c_str());
+        goal_handle->canceled(result);
+        return;
       }
 
-      // Log a message every second
-      current_time = std::chrono::system_clock::now();
-      if (current_time - time_since_msg >= 1s) {
-        RCLCPP_INFO(node_->get_logger(), "%s running...", taskName_.c_str());
-        time_since_msg = std::chrono::system_clock::now();
-      }
+      // TODO(orduno) Handle or reject an attempted pre-emption
 
-      status = onCycleUpdate(result);
+      auto status = onCycleUpdate();
```

Comment:
I'm generally not a fan of switch statements, but I think this is a good use.

```
switch(onCycleUpdate())
{
  case (Status::SUCCEEDED): ...
  case (Status::FAILED): ...
}


```

Reply:
Yeah, good catch